### PR TITLE
Add Reentrantlock to Standard Library

### LINKED ISSUE #2177: Backend: Concurrency: Write a namespace that wraps `ReentrantLock`.

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #2177

### CODE REVIEW

File: main/src/library/ReentrantLock.flix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// A wrapper around a Java ReentrantLock.
+///
+pub opaque type ReentrantLock = ##java.util.concurrent.locks.ReentrantLock
+
+// TODO: waiting for Condition.flix to be merged into master
```
Comment (magnus-madsen) [inline main/src/library/ReentrantLock.flix:22]: Done!

(Remember to update the two comments in Condition.flix)

File: main/src/library/ReentrantLock.flix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+opaque type Condition = ##java.util.concurrent.locks.Condition
+
+namespace ReentrantLock {
+
+    ///
+    /// Creates an instance of ReentrantLock with the given fairness policy.
+    /// `fair = False` is the default for Java.
+    ///
+    pub def makeLock(fair: Bool): ReentrantLock & Impure =
```
Comment (magnus-madsen) [inline main/src/library/ReentrantLock.flix:31]: In Flix, we tend to call these `new`, thus it becomes `ReetrantLock.new()` which locks nice, imo.

File: main/src/ca/uwaterloo/flix/tools/BenchmarkCompiler.scala [hunk 3]
```diff
@@ -185,6 +185,7 @@ object BenchmarkCompiler {
     flix.addInput("TestSet.flix", LocalResource.get("/test/ca/uwaterloo/flix/library/TestSet.flix"))
     flix.addInput("TestString.flix", LocalResource.get("/test/ca/uwaterloo/flix/library/TestString.flix"))
     flix.addInput("TestValidation.flix", LocalResource.get("/test/ca/uwaterloo/flix/library/TestValidation.flix"))
+    // TODO: add Condition.flix and ReentrantLock.flix here?
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/tools/BenchmarkCompiler.scala:188]: ?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/tools/BenchmarkCompiler.scala:188]: No, this is for tests?

File: main/src/library/Condition.flix [hunk 4]
```diff
@@ -40,7 +44,9 @@ namespace Condition {
         }
 
     ///
-    /// Causes the current thread to wait until it is signalled or interrupted, or the specified waiting time (nanos) elapses.
+    /// Causes the current thread to wait until it is signalled or interrupted, or the specified waiting
+    /// time(nanos) elapses. The method returns an estimate of the number of nanoseconds remaining to wait
+    /// given the supplied nanosTimeout value upon return, or a value less than or equal to zero if it timed out.
```
Comment (mlutze) [inline main/src/library/Condition.flix:49]: Would it be clearer to return an enum from this? There seem to be two categories of return value: `Remaining(Int64)` and `TimedOut`.

Comment (JonathanStarup) [reply main/src/library/Condition.flix:49]: Possibly yes, the java class also has a simpler method which just returns a bool which is the same concept. I haven't used this for anything myself so I'm not sure what the best is.

Comment (magnus-madsen) [reply main/src/library/Condition.flix:49]: I am going to leave this as future work.

**Pull request merged**